app/db/interface.py

```diff
@@ -156,33 +158,50 @@ def _query_payments(self, payment_attributes):
         except exc.SQLAlchemyError:
             raise PaymentServiceQueryError('Could not retrieve payment items due to query error with given attributes')
 
-# UTILITY FUNCTIONS
-    def is_valid_put(self, existing_user_id, new_data):
+#######################
+## UTILITY FUNCTIONS ##
+#######################
+
+    def is_valid(self, new_payment, user_id=None):
         valid = False
-        valid_detail = False
-        try:
-            for key in self.NONUPDATABLE_PAYMENT_RREQUEST_FIELDS :
-                if key in new_data:
-                   raise DataValidationError('Invalid payment: body of request contained bad or no data')
-            #if existing_user_id != data['user_id']:
-            #    raise DataValidationError('Invalid payment: Changes to user_id field not allowed')
-            type = new_data['payment_type']
-            detail = new_data['details']
-            if bool(re.search(r'\d', detail['user_name'])) == False:
-                valid = True
-            if type == 'credit' or type == 'debit':
-                card_number = detail['card_number']
-                expires_date = detail['expires']
-                if bool(re.match('^[0-9]+$', card_number)) and (len(card_number) == 16):
-                    datetime.strptime(expires_date, '%m/%Y')
-                    valid_detail = True
-            elif new_data['is_linked'] != None and type == 'paypal':
-                valid_detail = True
-        except KeyError as e:
-            raise DataValidationError('Invalid payment: missing ' + e.args[0])
-        except TypeError as e:
-            raise DataValidationError('Invalid payment: body of request contained bad or no data')
-        return valid & valid_detail
+
+        if user_id and int(new_payment['user_id']) != user_id:
+            raise DataValidationError('Error: You cannot modify the field: user_id')
+
+        for key in self.NONUPDATABLE_PAYMENT_FIELDS :
+            if key in new_payment:
+               raise DataValidationError('Error: You cannot modify the field: %s' % key)
+
+        payment_type = new_payment['payment_type']
+        user_name = new_payment['details']['user_name']
+
+        if bool(re.match("^[A-Za-z\-' ]{0,50}$", user_name)):
+            
+            if payment_type == 'credit' or payment_type == 'debit':
+                card_number = new_payment['details']['card_number']
+                card_type = new_payment['details']['card_type']
+                expires = new_payment['details']['expires']
+                
+                if bool(re.match('^[0-9]{16}$', card_number)):
+                    
+                    if bool(re.match("^[A-Za-z ]{0,10}$", card_type)):
+                        
+                        if bool(re.match('^[0-9]{2}/[0-9]{4}$', expires)):
```

Comment:
Hmm, isn't it possible to do an `AND` over several different regex groups in a regex?  I'm not a regex wizard but I think it should be an option.